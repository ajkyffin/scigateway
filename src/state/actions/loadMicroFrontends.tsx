/* eslint-disable @typescript-eslint/explicit-function-return-type */
// single-spa doesn't come with any types - all single-spa code should be limited to this file.
import * as singleSpa from 'single-spa';
import { Plugin } from '../state.types';

const runScript = async (url: string) => {
  return new Promise((resolve, reject) => {
    const script = document.createElement('script');
    script.src = url;
    script.onload = resolve;
    script.onerror = reject;

    const body = document.getElementsByTagName('body')[0];
    body.appendChild(script);
  });
};

const loadReactApp = async (name: string, url: string) => {
  await runScript(url);
  // Plugins are loaded on to the window and define their own property name so can't guess this at compile time
  /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
  return (window as any)[name];
};

async function loadApp(name: string, appURL: string) {
  // register the app with singleSPA and pass a reference to the store of the app as well as a reference to the globalEventDistributor
  singleSpa.registerApplication(
    name,
    () => loadReactApp(name, appURL),
    (location: URL) => {
      return location.pathname === '/';
    }
  );
}

async function init(plugins: Plugin[]) {
  /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
  const loadingPromises: Promise<any>[] = [];

  plugins
    .filter(p => p.enable)
    .forEach(p => {
      loadingPromises.push(loadApp(p.name, p.src));
    });

  // wait until all stores are loaded and all apps are registered with singleSpa
  await Promise.all(loadingPromises);

  singleSpa.start();
}

export default {
  init,
};
