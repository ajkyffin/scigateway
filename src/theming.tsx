import { createMuiTheme } from '@material-ui/core/styles';
import { ThemeOptions, Theme } from '@material-ui/core/styles/createMuiTheme';

export interface UKRIThemeOptions extends ThemeOptions {
  ukri: {
    purple: string;
    blue: string;
    orange: string;
    green: string;
    grey: string;
  };
}

export interface UKRITheme extends Theme {
  ukri: {
    purple: string;
    blue: string;
    orange: string;
    green: string;
    grey: string;
  };
}

export const buildTheme = (): Theme => {
  const options: UKRIThemeOptions = {
    typography: {
      useNextVariants: true,
    },
    palette: {
      primary: {
        main: '#1D4F91', // dark blue
      },
      secondary: {
        main: '#63666A', //grey
      },
    },
    ukri: {
      purple: '#8C4799',
      blue: '#1D4F91',
      orange: '#C34613',
      green: '#008275',
      grey: '#63666A',
    },
  };

  return createMuiTheme(options);
};
