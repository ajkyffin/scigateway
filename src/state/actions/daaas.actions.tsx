import axios from 'axios';
import { Action, AnyAction } from 'redux';
import log from 'loglevel';
import {
  ConfigureStringsType,
  ConfigureStringsPayload,
  ToggleDrawerType,
  AuthFailureType,
  AuthorisedPayload,
  AuthSuccessType,
  ApplicationStrings,
  SignOutType,
  FeatureSwitchesPayload,
  ConfigureFeatureSwitchesType,
  FeatureSwitches,
  LoadingAuthType,
  DismissNotificationType,
  DismissNotificationPayload,
} from '../daaas.types';
import { ActionType, ThunkResult, StateType } from '../state.types';
import loadMicroFrontends from './loadMicroFrontends';
import { push } from 'connected-react-router';
import { ThunkAction } from 'redux-thunk';

export const configureStrings = (
  appStrings: ApplicationStrings
): ActionType<ConfigureStringsPayload> => ({
  type: ConfigureStringsType,
  payload: {
    res: appStrings,
  },
});

export const loadStrings = (path: string): ThunkResult<Promise<void>> => {
  return async dispatch => {
    await axios
      .get(path)
      .then(res => {
        dispatch(configureStrings(res.data));
      })
      .catch(error =>
        log.error(`Failed to read strings from ${path}: ${error}`)
      );
  };
};

export const loadFeatureSwitches = (
  featureSwitches: FeatureSwitches
): ActionType<FeatureSwitchesPayload> => ({
  type: ConfigureFeatureSwitchesType,
  payload: {
    switches: featureSwitches,
  },
});

export const configureSite = (): ThunkResult<Promise<void>> => {
  return async dispatch => {
    await axios.get(`/settings.json`).then(res => {
      const settings = res.data;
      if (settings['features']) {
        dispatch(loadFeatureSwitches(settings['features']));
      }

      const uiStringResourcesPath = !settings['ui-strings'].startsWith('/')
        ? '/' + settings['ui-strings']
        : settings['ui-strings'];
      dispatch(loadStrings(uiStringResourcesPath));
      loadMicroFrontends.init(settings.plugins);
    });
  };
};

export const toggleDrawer = (): Action => ({
  type: ToggleDrawerType,
});

export const signOut = (): ThunkAction<
  void,
  StateType,
  null,
  AnyAction
> => dispatch => {
  dispatch({ type: SignOutType });
  dispatch(push('/'));
};

export const unauthorised = (): Action => ({
  type: AuthFailureType,
});

export const authorised = (token: string): ActionType<AuthorisedPayload> => ({
  type: AuthSuccessType,
  payload: {
    token,
  },
});

export const loadingAuthentication = (): Action => ({
  type: LoadingAuthType,
});

export const verifyUsernameAndPassword = (
  username: string,
  password: string
): ThunkResult<Promise<void>> => {
  return async (dispatch, getState) => {
    // will be replaced with call to login API for authentification
    dispatch(loadingAuthentication());
    await new Promise(resolve => setTimeout(resolve, 3000)).then(() => {
      if (username === 'INVALID_NAME' && password != null) {
        dispatch(unauthorised());
      } else {
        const token = 'validLoginToken';
        dispatch(authorised(token));

        // redirect the user to the original page they were trying to get to
        // the referrer is added by the redirect in routing.component.tsx
        const previousRouteState = getState().router.location.state;
        dispatch(
          push(
            previousRouteState && previousRouteState.referrer
              ? previousRouteState.referrer
              : '/'
          )
        );
      }
    });
  };
};

export const showDropdown = (): Action => ({
  type: ToggleDrawerType,
});

export const dismissMenuItem = (
  index: number
): ActionType<DismissNotificationPayload> => {
  return {
    type: DismissNotificationType,
    payload: {
      index,
    },
  };
};
