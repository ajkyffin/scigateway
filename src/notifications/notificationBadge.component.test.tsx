import React from 'react';
import { act } from 'react-dom/test-utils';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import NotificationBadge, {
  NotificationBadgeWithStyles,
} from './notificationBadge.component';
import Badge from '@material-ui/core/Badge';
import { MuiThemeProvider } from '@material-ui/core';
import { buildTheme } from '../theming';
import { Action } from 'redux';
import configureStore from 'redux-mock-store';
import { initialState } from '../state/reducers/scigateway.reducer';
import { dismissMenuItem } from '../state/actions/scigateway.actions';

describe('Notification Badge component', () => {
  let shallow;
  let mount;
  let mockStore;

  beforeEach(() => {
    shallow = createShallow({});
    mount = createMount();
    mockStore = configureStore();
  });

  const theme = buildTheme();

  it('Notification badge renders correctly', () => {
    const wrapper = shallow(
      <MuiThemeProvider theme={theme}>
        <NotificationBadgeWithStyles
          notifications={[{ message: 'my message', severity: 'warning' }]}
          deleteMenuItem={(): Action => ({ type: test })}
        />
      </MuiThemeProvider>
    );

    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

  it('renders correct number of notifications in the badge', () => {
    let wrapper = shallow(
      <MuiThemeProvider theme={theme}>
        <NotificationBadgeWithStyles
          notifications={[
            { message: 'my message', severity: 'warning' },
            { message: 'my other message', severity: 'success' },
          ]}
          deleteMenuItem={(): Action => ({ type: test })}
        />
      </MuiThemeProvider>
    );

    expect(
      wrapper
        .dive()
        .dive()
        .find(Badge)
        .prop('badgeContent')
    ).toEqual(2);

    wrapper = shallow(
      <MuiThemeProvider theme={theme}>
        <NotificationBadgeWithStyles
          deleteMenuItem={(): Action => ({ type: test })}
        />
      </MuiThemeProvider>
    );

    expect(
      wrapper
        .dive()
        .dive()
        .find(Badge)
        .prop('badgeContent')
    ).toBeNull();
  });

  it('sends dismissMenuItem action when dismissNotification prop is called', () => {
    let state = { scigateway: initialState };
    state.scigateway.notifications = [
      { message: 'my message', severity: 'warning' },
    ];
    const testStore = mockStore(state);

    const wrapper = shallow(
      <MuiThemeProvider theme={theme}>
        <NotificationBadge store={testStore} />
      </MuiThemeProvider>
    );

    wrapper
      .dive()
      .dive()
      .dive()
      .find('#notifications-menu [dismissNotification]')
      .prop('dismissNotification')();

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(dismissMenuItem(0));
  });

  it('opens menu when button clicked and closes menu when there are no more notifications', () => {
    //TODO: when enzyme supports hooks, change test to shallow and test the menuAnchor state
    let props = {
      notifications: [{ message: 'my message', severity: 'warning' }],
      deleteMenuItem: (): Action => ({ type: test }),
    };

    const wrapper = mount(
      React.createElement(
        props => (
          <MuiThemeProvider theme={theme}>
            <NotificationBadgeWithStyles {...props} />
          </MuiThemeProvider>
        ),
        props
      )
    );

    act(() => {
      wrapper
        .find('.NotificationBadge-button-1')
        .first()
        .prop('onClick')(
        wrapper
          .find('.NotificationBadge-button-1')
          .first()
          .getDOMNode()
      );
    });

    expect(wrapper.find('#notifications-menu').exists()).toBeTruthy();

    wrapper.setProps({ notifications: [] });

    expect(wrapper.find('#notifications-menu').exists()).toBeFalsy();
  });
});
