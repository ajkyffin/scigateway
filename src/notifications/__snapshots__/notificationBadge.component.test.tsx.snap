// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification Badge component Notification badge renders correctly 1`] = `
<div
  className="NotificationBadge-root-1"
>
  <WithStyles(IconButton)
    className="NotificationBadge-button-2"
    onClick={[Function]}
  >
    <WithStyles(Badge)
      badgeContent={1}
      classes={
        Object {
          "colorPrimary": "NotificationBadge-badge-3",
        }
      }
      color="primary"
    >
      <pure(NotificationsIcon) />
    </WithStyles(Badge)>
  </WithStyles(IconButton)>
  <WithStyles(Menu)
    anchorEl={null}
    id="notifications-menu"
    onClose={[Function]}
    open={false}
  >
    <WithStyles(DaaasNotification)
      dismissNotification={[Function]}
      index={0}
      key="0"
      message="my message"
      severity="warning"
    />
  </WithStyles(Menu)>
</div>
`;
